### PR TITLE
Lookup transform plugin

### PR DESCRIPTION
**Pull request by ailegion**
added[WIP]: Lookup transform plugin

JIRA: https://issues.cask.co/browse/PLUGIN-392

### CODE REVIEW

File: core-plugins/docs/Lookup-batchjoiner.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+Use Case
+--------
+This transform is used when you need to lookup a reference data set for finding a specific value or multiple 
+values for input data to complete the information before it is loaded into the data mart or warehouse. 
+This will help to provide complete information that needs to be loaded into the data mart.
+
+Properties
+----------
+**lookupDataset:** Amongst the inputs connected to the lookup transformation, this determines the input that should be
```
Comment (bdmogal) [inline core-plugins/docs/Lookup-batchjoiner.md]: Let's use the labels as they appear in the UI.

File: core-plugins/docs/Lookup-batchjoiner.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+|-------------|
+|id  |customer_id |phone_number |
+|1   |1           | 555-555-555 |
+|2   |1           | 333-333-333 |
+
+The output record will have the lookup value field which in our case is phone_number aliased as phone
+
+|customer|
+|--------|
+|id  |first_name  |last_name |phone_number |
```
Comment (bdmogal) [inline core-plugins/docs/Lookup-batchjoiner.md]: if it is aliased as `phone`, then this should be `phone`?

File: core-plugins/widgets/Lookup-batchjoiner.json [hunk 3]
```diff
@@ -0,0 +1,61 @@
+{
+  "metadata": {
+    "spec-version": "1.3"
```
Comment (bdmogal) [inline core-plugins/widgets/Lookup-batchjoiner.json]: 1.5

File: core-plugins/widgets/Lookup-batchjoiner.json [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  },
+  "inputs": {
+    "multipleInputs": true
+  },
+  "display-name" : "Lookup Transform",
```
Comment (bdmogal) [inline core-plugins/widgets/Lookup-batchjoiner.json]: only Lookup

Comment (ailegion) [reply core-plugins/widgets/Lookup-batchjoiner.json]: updated per PR review

File: core-plugins/docs/Lookup-batchjoiner.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+|customer_id |first_name |last_name |
+|1           |John       |Doe       |
+
+|phone_numbers|   |             |
+|----|------------|-------------|
+|id  |customer_id |phone_number |
+|1   |1           | 555-555-555 |
+|2   |1           | 333-333-333 |
+
+The output record will have the lookup value field which in our case is phone_number aliased as phone
```
Comment (rmstar) [inline core-plugins/docs/Lookup-batchjoiner.md:59]: Can you specify what these are set to:
- input key field
- lookup key field
- output field

File: core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Lookup.java [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+  public void configurePipeline(MultiInputPipelineConfigurer multiInputPipelineConfigurer) {
+    super.configurePipeline(multiInputPipelineConfigurer);
+    final Map<String, Schema> inputSchemas = multiInputPipelineConfigurer
+      .getMultiInputStageConfigurer().getInputSchemas();
+    FailureCollector collector = multiInputPipelineConfigurer.getMultiInputStageConfigurer().getFailureCollector();
+    config.validate(collector);
+    if (config.fieldsContainMacro()) {
+      return;
+    }
+    if (inputSchemas.values().size() != 2) {
```
Comment (rmstar) [inline core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Lookup.java:324]: Can you try testing the following:
- 1 input stage's output schema is a macro
- both input stages' output schemas are macros.

I think you'll hit an issue with the former case - please test and see if it works.

If you just need input stage names, you can use `getInputStages`. https://github.com/cdapio/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/MultiInputStageConfigurer.java#L44

**Review by rmstar — COMMENTED** [review]
Can you please add a unit test.

**Comment by ailegion** [comment]
> Can you please add a unit test.

fixed PR reviews and added unit test

**Pull request closed without merging**